daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala

```diff
@@ -765,19 +765,20 @@ private[lf] object SBuiltin {
     }
   }
 
-  /** $tproj[fieldIndex] :: Struct -> a */
-  final case class SBStructProj(fieldIndex: Int) extends SBuiltinPure(1) {
-    override private[speedy] def executePure(args: util.ArrayList[SValue]): SValue =
-      getSStruct(args, 0).values.get(fieldIndex)
-  }
-
   /** $tproj[field] :: Struct -> a */
-  // This is a slower version of `SBStructProj` for the case when we didn't run
-  // the Daml-LF type checker and hence didn't infer the field index.
-  final case class SBStructProjByName(field: Ast.FieldName) extends SBuiltinPure(1) {
+  final case class SBStructProj(field: Ast.FieldName) extends SBuiltinPure(1) {
+    // The variable `fieldIndex` is used to cache the (logarithmic) evaluation
+    // of `struct.fieldNames.indexOf(field)` at the first call in order to
+    // avoid its reevaluations, hence obtaining an amortized constant
+    // complexity.
+    // Here we take advantage of the Java memory model
+    // (https://docs.oracle.com/javase/specs/jls/se8/html/jls-17.html#jls-17.7)
```

Comment:
I don’t understand how word tearing comes into play here.

Reply:
suggestion```
```

Reply:
I am not sure what you mean here. 
I just delete the last sentence.

Reply:
The `CompiledPackage` can be used concurrently by two different threads (e.g. by the evaluation of two different API calls). So we need to be sure the variable assignment is atomic. This is similar to the cash we have inside `SEVal` for the which Martin asked me to write a similar comment.

Reply:
ah apologies, I got confused and thought you linked to the section above. I think 17.7 is still not quite the right section. It doesn’t directly say that updating an int is atomic, it says that updating a long is non-atomic but close enough.